app/bin/tools_backfill_scopes.dart

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:gcloud/db.dart';
+import 'package:pub_dartlang_org/models.dart';
+import 'package:pub_dartlang_org/backend.dart' show autodetectScopes;
+
+import 'tools_common.dart';
+
+Future main(List<String> args) async {
+  int updatedCount = 0;
+  await withProdServices(() async {
+    await for (PackageVersion pv in dbService.query(PackageVersion).run()) {
+      final List<String> scopes = autodetectScopes(pv.pubspec);
+      if (!_isSameScope(pv.scopes, scopes)) {
+        try {
+          await _updatePackageVersion(pv.key);
+          updatedCount++;
+        } catch (e) {
+          print(
+              'Failed to update package ${pv.package} / ${pv.version}, error: $e');
+        }
+      }
+    }
+  });
+  print('Updated: $updatedCount package versions.');
+}
+
+bool _isSameScope(List<String> a, List<String> b) {
+  final int lengthA = a?.length ?? 0;
+  final int lengthB = b?.length ?? 0;
+  if (lengthA != lengthB) return false;
+  if (lengthA == 0) return true;
+  return (a.toSet()..removeAll(b)).isEmpty;
+}
+
+Future _updatePackageVersion(Key pvKey) async {
+  await dbService.withTransaction((Transaction t) async {
+    final list = (await t.lookup([pvKey.parent, pvKey]));
```

Comment:
I think simplicity of the code is more beneficial here than the one-off gain in performance. If/when we introduce more search features, we may need to modify this backfill job and make it more complex. I'd defer this optimization until then.

In addition to that: we could choose an indexing/search mechanism that doesn't require the `scope` field on the Package object, and we could remove the lookup here.